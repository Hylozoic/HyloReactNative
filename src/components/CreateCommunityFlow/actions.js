export const MODULE_NAME = `CreateCommunityFlow`
// export const ADD_COMMUNITY_NAME = `${MODULE_NAME}/ADD_COMMUNITY_NAME`
// export const ADD_COMMUNITY_DOMAIN = `${MODULE_NAME}/ADD_COMMUNITY_DOMAIN`
// export const ADD_COMMUNITY_PRIVACY = `${MODULE_NAME}/ADD_COMMUNITY_PRIVACY`

export const FETCH_COMMUNITY_EXISTS = `${MODULE_NAME}/FETCH_URL_EXISTS`

export function fetchCommunityExists (slug) {
  return {
    type: FETCH_COMMUNITY_EXISTS,
    graphql: {
      query: `
        query ($slug: String) {
          communityExists (slug: $slug) {
            exists
          }
        }
      `,
      variables: {
        slug
      }
    }
  }
}
