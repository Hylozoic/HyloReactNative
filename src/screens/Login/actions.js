import { clearSessionCookie } from 'util/session'
import { LoginManager } from 'react-native-fbsdk'
import { GoogleSignin } from '@react-native-community/google-signin'
import { isEmpty } from 'lodash'

export const LOGIN = 'LOGIN'
export const LOGIN_BY_TOKEN = 'LOGIN_BY_TOKEN'
export const LOGIN_WITH_APPLE = 'LOGIN_WITH_APPLE'
export const LOGIN_WITH_FACEBOOK = 'LOGIN_WITH_FACEBOOK'
export const LOGIN_WITH_GOOGLE = 'LOGIN_WITH_GOOGLE'
export const LOGOUT = 'LOGOUT'

export function login (email, password) {
  return {
    type: LOGIN,
    payload: {
      api: { method: 'post', path: '/noo/login', params: { email, password } }
    },
    meta: { email }
  }
}

export function loginByToken (userId, loginToken) {
  return {
    type: LOGIN_BY_TOKEN,
    payload: {
      api: { method: 'post', path: '/noo/login/token', params: { u: userId, t: loginToken } }
    }
  }
}

export function loginWithApple (params) {
  return {
    type: LOGIN_WITH_APPLE,
    payload: {
      api: { method: 'post', path: '/noo/login/apple/oauth', params }
    }
  }
}

export function loginWithFacebook (accessToken) {
  return {
    type: LOGIN_WITH_FACEBOOK,
    payload: {
      api: { method: 'post', path: `/noo/login/facebook-token/oauth?access_token=${accessToken}` }
    }
  }
}

export function loginWithGoogle (accessToken) {
  return {
    type: LOGIN_WITH_GOOGLE,
    payload: {
      api: { method: 'post', path: `/noo/login/google-token/oauth?access_token=${accessToken}` }
    }
  }
}

export function logout () {
  return {
    type: LOGOUT,
    payload: {
      api: {
        method: 'delete',
        path: '/noo/session',
        transform: () =>
          clearSessionCookie()
            .then(() =>
              LoginManager.logOut()
            )
            // TODO: This should be already handled by the same code in login/actions/Logout
            .then(async () => {
              if (!isEmpty(await GoogleSignin.getCurrentUser())) {
                return GoogleSignin.signOut()
              }
            })
      }
    }
  }
}
