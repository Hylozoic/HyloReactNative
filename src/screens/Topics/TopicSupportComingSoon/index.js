import component from './TopicSupportComingSoon'

export default component
