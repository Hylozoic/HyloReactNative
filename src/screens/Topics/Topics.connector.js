import { connect } from 'react-redux'
import getCurrentGroup from 'store/selectors/getCurrentGroup'
import fetchGroupTopics, { FETCH_GROUP_TOPICS } from 'store/actions/fetchGroupTopics'
import {
  getGroupTopics, presentGroupTopic, setTopicSubscribe, getTerm, setTerm
} from './Topics.store'
import { get } from 'lodash/fp'

export function mapStateToProps (state, props) {
  const searchTerm = getTerm(state)
  const group = getCurrentGroup(state, props)
  const pending = state.pending[FETCH_GROUP_TOPICS]
  const queryResultParams = {
    id: get('id', group),
    autocomplete: ''
  }
  const topicFilter = ct => ct.topic.name.toLowerCase().indexOf(searchTerm?.toLowerCase()) > -1
  const topicSort = (a, b) => {
    if (a.isSubscribed && !b.isSubscribed) return -1
    if (!a.isSubscribed && b.isSubscribed) return 1
    return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
  }
  const allTopics = getGroupTopics(state, queryResultParams)
  const groupHasTopics = allTopics.length > 0
  const filteredTopics = allTopics
    .filter(topicFilter)
    .map(presentGroupTopic)
    .sort(topicSort)

  return {
    pending,
    group,
    groupHasTopics,
    filteredTopics,
    searchTerm
  }
}

export const mapDispatchToProps = {
  fetchGroupTopics,
  setTopicSubscribe,
  setTerm
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { group } = stateProps
  const { navigation } = ownProps
  const groupId = group?.id
  const fetchGroupTopics = () =>
    dispatchProps.fetchGroupTopics(groupId, { first: null })
  const setTopicSubscribe = (topicId, isSubscribing) =>
    dispatchProps.setTopicSubscribe(topicId, groupId, isSubscribing)
  const goToTopic = topicName => navigation.navigate('Topic Feed', { topicName })

  return {
    ...ownProps,
    ...stateProps,
    ...dispatchProps,
    fetchGroupTopics,
    setTopicSubscribe,
    goToTopic
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
