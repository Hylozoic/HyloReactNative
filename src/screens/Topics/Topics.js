import React, { useRef, useCallback } from 'react'
import { Text, ScrollView, Image, View, TouchableOpacity } from 'react-native'
import { useFocusEffect, useIsFocused, useScrollToTop } from '@react-navigation/native'
import KeyboardFriendlyView from 'components/KeyboardFriendlyView'
import Loading from 'components/Loading'
import StarIcon from 'components/StarIcon'
import Badge from 'components/Badge'
import SearchBar from 'components/SearchBar'
import styles from './Topics.styles'
import EntypoIcon from 'react-native-vector-icons/Entypo'
import LinearGradient from 'react-native-linear-gradient'
import { isEmpty, get } from 'lodash/fp'
import { bannerlinearGradientColors } from 'style/colors'

export default function Topics ({
  groupHasTopics,
  group,
  filteredTopics,
  pending,
  setTopicSubscribe,
  goToTopic,
  searchTerm,
  setTerm,
  fetchGroupTopics
}) {
  const ref = useRef(null)

  useScrollToTop(ref)
  useFocusEffect(
    useCallback(() => {
      group?.id && fetchGroupTopics()
    }, [group.id])
  )

  const bannerUrl = get('bannerUrl', group)
  const name = get('name', group)
  const image = { uri: bannerUrl }

  return (
    <KeyboardFriendlyView style={styles.container}>
      <ScrollView ref={ref}>
        <View style={styles.bannerContainer}>
          {image && <Image source={image} style={styles.image} />}
          <LinearGradient style={styles.gradient} colors={bannerlinearGradientColors} />
          <View style={styles.titleRow}>
            <Text style={styles.name}>{name}</Text>
          </View>
        </View>
        <SearchBar
          style={styles.searchBar}
          value={searchTerm}
          onChangeText={setTerm}
          placeholder='Search Topics'
          disable={!groupHasTopics}
        />
        {pending && isEmpty(filteredTopics)
          ? <Loading />
          : <TopicList
              groupHasTopics={groupHasTopics}
              searchTerm={searchTerm}
              filteredTopics={filteredTopics}
              setTopicSubscribe={setTopicSubscribe}
              goToTopic={goToTopic}
            />}
      </ScrollView>
    </KeyboardFriendlyView>
  )
}

export function TopicList ({
  groupHasTopics,
  filteredTopics,
  setTopicSubscribe,
  searchTerm,
  goToTopic
}) {
  if (!groupHasTopics) return (
    <View style={styles.topicList}>
      <Text style={styles.emptyList}>No topics were found for this group</Text>
    </View>
  )
  return (
    <View style={styles.topicList}>
      {isEmpty(filteredTopics) && searchTerm && (
        <Text style={styles.emptyList}>No topics matched "{searchTerm}"</Text>
      )}
      {!isEmpty(filteredTopics) && filteredTopics.map((topic, i) => (
        <TopicRow
          topic={topic} key={i} first={i === 0}
          setTopicSubscribe={setTopicSubscribe}
          goToTopic={goToTopic}
        />
      ))}
    </View>
  )
}

export function TopicRow ({ topic, first, setTopicSubscribe, goToTopic }) {
  const { name, newPostCount, isSubscribed } = topic
  return (
    <TouchableOpacity
      style={[styles.topicRow, first && styles.firstRow]}
      onPress={() => goToTopic(topic.name)}
    >
      <SubscribeStar isSubscribed={isSubscribed} onPress={() => setTopicSubscribe(topic.id, !isSubscribed)} />
      <Text style={styles.topicName}>#{name}</Text>
      <View style={styles.rightItems}>
        <Badge style={styles.badge} count={newPostCount} />
        <EntypoIcon style={styles.chevron} name='chevron-right' />
      </View>
    </TouchableOpacity>
  )
}

export function SubscribeStar ({ isSubscribed, onPress }) {
  const theme = isSubscribed
    ? styles.subscribedStar
    : styles.unSubscribedStar
  return (
    <TouchableOpacity
      style={styles.star}
      onPress={() => onPress(!isSubscribed)}
      hitSlop={{ top: 10, bottom: 10, left: 15, right: 15 }}
    >
      <StarIcon theme={theme} />
    </TouchableOpacity>
  )
}
