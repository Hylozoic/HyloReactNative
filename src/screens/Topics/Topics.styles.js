import { StyleSheet } from 'react-native'
import { white, caribbeanGreen, capeCod40, capeCod20, rhino50, ghost } from 'style/colors'
import { isIOS } from 'util/platform'

const hasTextShadow = {
  textShadowColor: 'rgba(0, 0, 0, 0.25)',
  textShadowOffset: { width: 0, height: 2 },
  textShadowRadius: 7
}

export default {
  container: {
    backgroundColor: white,
    flex: 1
  },
  bannerContainer: {
    zIndex: 10,
    height: 142,
    marginBottom: 10
  },
  image: {
    height: 142,
    width: '100%',
    position: 'absolute'
  },
  gradient: {
    height: 142,
    width: '100%',
    position: 'absolute'
  },
  titleRow: {
    position: 'absolute',
    left: 0,
    bottom: 56,
    marginHorizontal: 16,
    flexDirection: 'row',
    alignItems: 'flex-end'
  },
  name: {
    fontSize: 24,
    fontFamily: 'Circular-Black',
    color: white,
    backgroundColor: 'transparent',
    ...hasTextShadow
  },
  // 
  searchBar: {
    marginHorizontal: 8,
    marginBottom: 12
  },
  topicList: {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderTopColor: capeCod40,
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: capeCod40,
    paddingLeft: 15
  },
  emptyList: {
    paddingVertical: 10
  },
  topicRow: {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderTopColor: capeCod40,
    height: 44,
    flexDirection: 'row',
    alignItems: 'center'
  },
  firstRow: {
    borderTopWidth: 0
  },
  topicName: {
    fontSize: 16
  },
  rightItems: {
    flexDirection: 'row',
    marginLeft: 'auto'
  },
  badge: {
    marginRight: 10,
    position: 'relative',
    top: 3
  },
  chevron: {
    marginRight: 12,
    fontSize: 24,
    color: capeCod20
  },
  star: {
    marginRight: 15
  },
  // themes
  unSubscribedStar: {
    icon: {
      fontSize: 30,
      color: capeCod20,
      backgroundColor: 'white'
    }
  },
  subscribedStar: {
    icon: {
      fontSize: 30,
      color: caribbeanGreen,
      backgroundColor: 'transparent'
    }
  }
}
