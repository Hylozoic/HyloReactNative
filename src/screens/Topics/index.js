import component from './Topics'
import connector from './Topics.connector'

export default connector(component)
