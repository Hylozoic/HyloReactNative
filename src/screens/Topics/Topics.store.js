import { makeGetQueryResults, makeQueryResultsModelSelector } from 'store/reducers/queryResults'
import { FETCH_GROUP_TOPICS } from 'store/actions/fetchGroupTopics'
export { setTopicSubscribe } from 'screens/Feed/Feed.store'

export const MODULE_NAME = 'Topics'
export const SET_TERM = `${MODULE_NAME}/SET_TERM`

const getGroupTopicResults = makeGetQueryResults(FETCH_GROUP_TOPICS)

export const getGroupTopics = makeQueryResultsModelSelector(
  getGroupTopicResults,
  'GroupTopic')

export function presentGroupTopic (groupTopic) {
  return {
    ...groupTopic.topic.ref,
    isSubscribed: groupTopic.isSubscribed,
    newPostCount: groupTopic.newPostCount
  }
}

export const defaultState = {
  term: ''
}

export default function reducer (state = defaultState, action) {
  const { error, type, payload } = action
  if (error) return state

  switch (type) {
    case SET_TERM:
      return {
        ...state,
        term: payload
      }
    default:
      return state
  }
}

export function setTerm (term) {
  return {
    type: SET_TERM,
    payload: term
  }
}

export function getTerm (state) {
  return state[MODULE_NAME].term
}
