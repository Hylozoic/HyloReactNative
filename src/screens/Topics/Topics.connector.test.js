import { mapStateToProps, mapDispatchToProps, mergeProps } from './Topics.connector'
import { MODULE_NAME } from './Topics.store'
import { FETCH_GROUP_TOPICS } from 'store/actions/fetchGroupTopics'

describe('mapStateToProps', () => {
  it('returns the right keys', () => {
    const state = {
      pending: {
        [FETCH_GROUP_TOPICS]: true
      },
      queryResults: {},
      [MODULE_NAME]: {}
    }
    expect(mapStateToProps(state, {})).toMatchSnapshot()
  })
})

describe('mapDispatchToProps', () => {
  it('returns the right keys', () => {
    expect(mapDispatchToProps).toMatchSnapshot()
  })
})

describe('mergeProps', () => {
  it('sets up functions', () => {
    const groupId = 132
    const topicId = 444
    const isSubscribing = true
    const topicName = 'mice'
    const stateProps = {
      group: {
        id: groupId
      }
    }
    const dispatchProps = {
      fetchGroupTopics: jest.fn(),
      setTopicSubscribe: jest.fn()
    }
    const ownProps = {
      navigation: {
        navigate: jest.fn()
      }
    }
    const mergedProps = mergeProps(stateProps, dispatchProps, ownProps)
    expect(mergedProps).toMatchSnapshot()
    mergedProps.fetchGroupTopics()
    expect(dispatchProps.fetchGroupTopics).toHaveBeenCalledWith(groupId, { first: null })
    mergedProps.setTopicSubscribe(topicId, isSubscribing)
    expect(dispatchProps.setTopicSubscribe).toHaveBeenCalledWith(topicId, groupId, isSubscribing)
    mergedProps.goToTopic(topicName)
    expect(ownProps.navigation.navigate).toHaveBeenCalledWith('Topic Feed', { topicName: topicName })
  })
})
