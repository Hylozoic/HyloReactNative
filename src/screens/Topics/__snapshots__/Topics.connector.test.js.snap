// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps returns the right keys 1`] = `
Object {
  "fetchCommunityTopics": [Function],
  "selectCommunity": [Function],
  "setTerm": [Function],
  "setTopicSubscribe": [Function],
}
`;

exports[`mapStateToProps returns the right keys 1`] = `
Object {
  "community": undefined,
  "communityHasTopics": false,
  "filteredTopics": Array [],
  "network": undefined,
  "pending": true,
  "searchTerm": undefined,
}
`;

exports[`mergeProps sets up functions 1`] = `
Object {
  "community": Object {
    "id": 132,
  },
  "fetchCommunityTopics": [Function],
  "goToTopic": [Function],
  "navigation": Object {
    "navigate": [MockFunction],
  },
  "setTopicSubscribe": [Function],
}
`;
